src/bgem/bspline/bspline_approx.py

```diff
@@ -595,37 +992,50 @@ def _compute_errors(self, point_loc, z_vec):
         n_patches = (self._u_basis.size - 2) * (self._v_basis.size - 2)
         u_n_basf = self._u_basis.size
         v_n_basf = self._v_basis.size
-
+        n_points_glob = 0
         g_vec = self._z_quad_points[:]
         n = g_vec.shape[0]
         err = np.zeros([n])
 
         linsp = np.array([0, 1, 2], dtype=int)
         linsp31 = np.repeat(linsp, 3)  # linsp_v
         linsp13 = np.tile(linsp, 3)  # linsp_u
+        err_mat_max = np.zeros((self._u_basis.size - 2, self._v_basis.size - 2))
+        err_mat_eucl2 = np.zeros((self._u_basis.size - 2, self._v_basis.size - 2))
 
         for interval_id in range(0, n_patches):
             n_loc_points = len(point_loc[interval_id])
             n_uv_loc_nz = (self._u_basis.degree + 1) * (self._v_basis.degree + 1)
             row = np.zeros(n_loc_points * n_uv_loc_nz, dtype=int)
             col = np.zeros(n_loc_points * n_uv_loc_nz, dtype=int)
             data = np.zeros(n_loc_points * n_uv_loc_nz)
-            nnz_b = 0
-            for idx in point_loc[interval_id]:
+            datad = np.zeros([n_loc_points, n_uv_loc_nz])
+            if len(point_loc[interval_id]) > 0:
+                idx = point_loc[interval_id][0]
                 u, v = self._uv_quad_points[idx, 0:2]
                 iu = self._u_basis.find_knot_interval(u)
                 iv = self._v_basis.find_knot_interval(v)
-                u_base_vec = self._u_basis.eval_base_vector(iu, u)
-                v_base_vec = self._v_basis.eval_base_vector(iv, v)
-                data[nnz_b:nnz_b + 9] = np.kron(v_base_vec, u_base_vec)
-                col[nnz_b: nnz_b + 9] = (linsp31 + iv) * u_n_basf + iu + linsp13
-                row[nnz_b:nnz_b + 9] = idx
-                nnz_b += 9
-            if len(point_loc[interval_id]) > 0:
-                mat_b = scipy.sparse.csr_matrix((data, (row, col)), shape=(n, u_n_basf * v_n_basf))
-                err += mat_b.dot(z_vec)
-        err = err - g_vec
-        return err
+                col = (linsp31 + iv) * u_n_basf + iu + linsp13
+                z_loc = z_vec[col]
+                #nnz_b = 0
+                n_points = 0
+                patch_point_loc = point_loc[interval_id]
+                for idx in patch_point_loc:
```

Comment:
We can modify the eval_base_vector function to accept vector of `u` values and return a matrix 3xN. Then this loop can be rewritten using the matrix NumPy operations. That can provide a huge speedup and is good step even if the basis would be rewritten using C++.

If this shows effect, we can try to modify matrix assembly in a similar way.